server/src/main/java/org/opensearch/common/util/concurrent/ThreadContext.java

```diff
@@ -230,7 +230,11 @@ public StoredContext stashAndMergeHeaders(Map<String, String> headers) {
      * @param preserveResponseHeaders if set to <code>true</code> the response headers of the restore thread will be preserved.
      */
     public StoredContext newStoredContext(boolean preserveResponseHeaders) {
-        return newStoredContext(preserveResponseHeaders, Collections.emptyList());
+        return newStoredContext(preserveResponseHeaders, Collections.emptyList(), false);
+    }
+
+    public StoredContext newStoredContext(boolean preserveResponseHeaders, boolean forIndependentTask) {
```

Comment:
@Gaganjuneja I believe adding more flags to the method is not an answer, the existing `newStoredContext` has the mechanism to no propagate specific transient headers which we could use:

```
ThreadContext.StoredContext storedContext = threadContext.newStoredContext(true, Collections.singletonList(CURRENT_SPA));
```

Reply:
@reta, Yes we can do that, Even I had tested like this only. But earlier when we introduced the propagators our thought process was that generic classes such as ThreadContext and ThreadPool shouldn't be aware of the telemetry specific attributes.

Reply:
@reta your thoughts here?

Reply:
> @reta your thoughts here?

@Gaganjuneja sorry haven't seem the comment, I think we don't need to expose the attributes if we make it the default for `newStoredContext`:

```
    public StoredContext newStoredContext(boolean preserveResponseHeaders) {
        return newStoredContext(preserveResponseHeaders, Collections.singletonList(CURRENT_SPAN));
    }
```

Reply:
> We need to propagate span details during all thread handoffs except the scheduler task.

Don't we use `stashContext` there?

Reply:
StashContext is happening after the newStoredContext.

https://github.com/opensearch-project/OpenSearch/blob/074bc6a91c05a5ed0e3d36cd672c905b0bf023c3/server/src/main/java/org/opensearch/common/util/concurrent/ThreadContext.java#L839

Reply:
StashContext is actually not serving our intended purpose because first it creates the newStoredContext from the calling thread's context and just before executing the run method, it does the stashContext then newStoredContext:restore and then execute the runnable.

Reply:
I think we are dealing with larger problem, here are few questions that bothers me at the moment:
 - Why periodic scheduled runnable inherit the context at all? Sometime it is needed (fe delayed computation) but in general - they should be executed in isolated context
 - We are focusing on tracing but the same issue applies to Task ID propagation, if something is scheduled within task scope, the task id will stuck as well (it just has no visible effects like traces do)

I will try to find more time to investigate all these different cases.

Reply:
> Why periodic scheduled runnables inherit the context at all? Sometimes it is needed (fe delayed computation) but in general - they should be executed in isolated context.

Agreed with you. I am not sure why this is implemented this way? But its clearly mentioned in the java oc.

https://github.com/opensearch-project/OpenSearch/blob/332f4c535643b459071aa9d2b3e5d50d137d8523/server/src/main/java/org/opensearch/threadpool/ThreadPool.java#L450

>We are focusing on tracing but the same issue applies to Task ID propagation, if something is scheduled within task scope, the task id will stuck as well (it just has no visible effects like traces do)

Task_ID is just being populated in the Search path for the Task resource tracking so might not be causing the issue.

Reply:
@reta your thoughts?

Reply:
@Gaganjuneja apologies for delays, holidays time, there are 2 directions which I am exploring right now:
 - `threadContext.markAsSystemContext();` this one seems to be overlooked so far BUT it is a core part of the most background sync actions, the hypothesis here is that the system context should never inherit the tracing context (as per documentation of `markAsSystemContext`)
 - introduce `ThreadContextPolicy` for `ThreadPool::schedule` to allow callers to manage context propagation in a structured way, the hypothesis here is that is not impossible to find out if context should or should not be propagated (delayed or isolated scheduled execution) and only caller knows that

Reply:
@reta,
> threadContext.markAsSystemContext(); this one seems to be overlooked so far BUT it is a core part of the most background sync actions, the hypothesis here is that the system context should never inherit the tracing context (as per documentation of markAsSystemContext)

markAsSystemContext makes sense, we can skip copying the context if isSystemContext is true. I will explore more to see where all this is being used.

Reply:
@reta, I have gone through and you are right, `markAsSystemContext` is being used mostly at the system actions level. It seems to be a right approach to call the clear method of propagators or simply clear the transient header in the `markAsSystemContext` call. your thoughts?

Reply:
@Gaganjuneja thank for confirming, yeah, it sounds reasonable, thank you.